src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -1434,3 +1434,103 @@ void QgsDataDefinedSizeLegendNode::cacheImage() const
   }
 }
 
+QgsVectorLabelLegendNode::QgsVectorLabelLegendNode( QgsLayerTreeLayer *nodeLayer, const QgsPalLayerSettings &labelSettings, QObject *parent ): QgsLayerTreeModelLegendNode( nodeLayer, parent ), mLabelSettings( labelSettings )
+{
+}
+
+QgsVectorLabelLegendNode::~QgsVectorLabelLegendNode()
+{
+}
+
+QVariant QgsVectorLabelLegendNode::data( int role ) const
+{
+  if ( role == Qt::DisplayRole )
+  {
+    return mUserLabel;
+  }
+  if ( role == Qt::DecorationRole )
+  {
+    const int iconSize = QgsLayerTreeModel::scaleIconSize( 16 );
+    return QgsPalLayerSettings::labelSettingsPreviewPixmap( mLabelSettings, QSize( iconSize, iconSize ), QStringLiteral( "Aa" ) );
+  }
+  return QVariant();
+}
+
+QSizeF QgsVectorLabelLegendNode::drawSymbol( const QgsLegendSettings &settings, ItemContext *ctx, double itemHeight ) const
+{
+  Q_UNUSED( itemHeight );
+  if ( !ctx )
+  {
+    return QSizeF( 0, 0 );
+  }
+
+  return drawSymbol( settings, ctx->painter, ctx->columnLeft, ctx->top );
+}
+
+QSizeF QgsVectorLabelLegendNode::drawSymbol( const QgsLegendSettings &settings, QPainter *p, double xOffset, double yOffset ) const
+{
+  QColor textColor;
+  QFont f = font( textColor );
+
+  double textHeight = settings.fontHeightCharacterMM( f, QChar( 'A' ) );
+  double textWidth = settings.textWidthMillimeters( f, "Aa" );
+
+  if ( p )
+  {
+    p->setPen( textColor );
+    settings.drawText( p, xOffset + settings.symbolSize().width() / 2.0 - textWidth / 2.0, yOffset + settings.symbolSize().height() / 2.0 + textHeight / 2.0, "Aa", f );
```

Comment:
This should be using QgsTextRenderer to render the text, so that buffers/shadows/backgrounds/etc are correctly rendered

Reply:
You are right, it is modified it to use QgsTextRenderer now.
Actually it would be good to reuse the code in QgsPalLayerSettings::labelSettingsPreviewPixmap by refactor it to a method QgsPalLayerSettings::labelSettingsPreview( QPainter* ) and use that to render both the pixmap and the legend node. Is it okay to do so or are there any objections?

Reply:
Oh, it seems QgsPalLayerSettings::labelSettingsPreview does a bit too much (e.g. draw a frame around the icon rectangle). So I'll keep using QgsTextRenderer directly